tests/unit/test_singleton_snap.py

```diff
@@ -0,0 +1,138 @@
+    p.start()
+
+    # Wait until the lock file appears to ensure the lock is held, up to 1 second.
+    lock_file = f'{lock_dir}/LCK..otelcol'
+    for _ in range(10):
+        if os.path.exists(lock_file):
+            break
+        time.sleep(0.1)
+    else:
+        pytest.fail('Lock file was not created in time')
+
+    mgr2 = SingletonSnapManager('unit-2')
+    start = time.time()
+    with pytest.raises(SingletonSnapError):
+        with mgr2.snap_operation('otelcol', timeout=1):
+            pass
+    assert time.time() - start >= 1
+
+    p.join()
+
+
+def try_snap_lock(unit, result_queue, lock_dir):
+    SingletonSnapManager.LOCK_DIR = lock_dir
+    mgr = SingletonSnapManager(unit)
+    # The timeout is set to 1 second but the operation sleeps for 2 seconds
+    # to ensure that only one process can acquire the lock.
+    try:
+        with mgr.snap_operation('otelcol', timeout=1):
+            result_queue.put(True)
+            time.sleep(2)
+    except SingletonSnapError:
+        result_queue.put(False)
+
+
+def test_exclusive_snap_operation(lock_dir):
+    acquired = Queue()
+    p1 = Process(target=try_snap_lock, args=('unit-1', acquired, lock_dir))
+    p2 = Process(target=try_snap_lock, args=('unit-2', acquired, lock_dir))
+
+    p1.start()
+
+    # Wait until the lock file appears to ensure the lock is held, up to 1 second.
+    lock_file = f'{lock_dir}/LCK..otelcol'
+    for _ in range(10):
+        if os.path.exists(lock_file):
+            break
+        time.sleep(0.1)
+    else:
+        pytest.fail('Lock file was not created in time')
+
+    p2.start()
+    p1.join()
+    p2.join()
+
+    results = [acquired.get(), acquired.get()]
+    assert results.count(True) == 1
+    assert results.count(False) == 1
+
+
+def try_config_lock(unit, result_queue, lock_dir):
+    SingletonSnapManager.LOCK_DIR = lock_dir
+    mgr = SingletonSnapManager(unit)
+    # The timeout is set to 1 second but the operation sleeps for 2 seconds
+    # to ensure that only one process can acquire the lock.
+    try:
+        with mgr.config_operation('/tmp/test.yaml', timeout=1):
+            result_queue.put(True)
+            time.sleep(2)
+    except SingletonSnapError:
+        result_queue.put(False)
+
+
+def test_exclusive_config_operation(lock_dir):
+    acquired = Queue()
+    p1 = Process(target=try_config_lock, args=('unit-1', acquired, lock_dir))
+    p2 = Process(target=try_config_lock, args=('unit-2', acquired, lock_dir))
+
+    p1.start()
+
+    # Wait until the lock file appears to ensure the lock is held, up to 1 second.
+    lock_file = f'{lock_dir}/LCK..config__tmp_test_yaml'
+    for _ in range(10):
+        if os.path.exists(lock_file):
+            break
+        time.sleep(0.1)
+    else:
+        pytest.fail('Lock file was not created in time')
+
+    p2.start()
+    p1.join()
+    p2.join()
+
+    results = [acquired.get(), acquired.get()]
+    assert results.count(True) == 1
+    assert results.count(False) == 1
```

Comment:
Should we scale up the concurrent test to e.g. 20 processes?

Reply:
Good idea, will do.